File: Difficulty: Medium/Search Pattern (KMP-Algorithm)/search-pattern-kmpalgorithm.py
#User function Template for python3

class Solution:
    def search(self, pat, txt):
        # code here
        # 02.12.24 POTD
        ans = []
        n, m = len(txt), len(pat)
        
        for i in range(n-m+1):
            if txt[i:i+m] == pat:
                ans.append(i)
                
        return ans


#{ 
 # Driver Code Starts
#Initial Template for Python 3

if __name__ == '__main__':
    t = int(input())
    for _ in range(t):
        s = input().strip()
        patt = input().strip()
        ob = Solution()
        ans = ob.search(patt, s)
        if len(ans) == 0:
            print("[]", end="")
        for value in ans:
            print(value, end=' ')
        print()

# } Driver Code Ends

File: Medium/Search Pattern (KMP-Algorithm)/search-pattern-kmpalgorithm.cpp
//{ Driver Code Starts
#include<bits/stdc++.h>
using namespace std;


// } Driver Code Ends
class Solution
{
    public:
        vector <int> search(string pat, string txt)
        {
            //code here
            // 09.01.24 POTd
            vector<int> ans;
            int p = pat.size();
            int lim = txt.size()-p;
            
            for(int i=0; i<=lim; i++){
                if(txt[i] == pat[0] && txt.substr(i,p) == pat){
                    ans.push_back(i+1);
                }
            }
            
            return (ans.empty() ? vector<int>{-1} : ans);
        }
     
};

//{ Driver Code Starts.
int main()
{
    int t;
    cin >> t;
    while (t--)
    {
        string S, pat;
        cin >> S >> pat;
        Solution ob;
        vector <int> res = ob.search(pat, S);
        if (res.size()==0)
            cout<<-1<<endl;
        else {
            for (int i : res) cout << i << " ";
            cout << endl;
        }
    }
    return 0;
}

// Contributed By: Pranay Bansal

// } Driver Code Ends